### PR TITLE
Refactor: Remove 6 unused tokenizer implementations and consolidate shared logic

### PR DESCRIPTION
**Pull request by fumiya-kume**
## Summary
- Remove 6 unused tokenizer implementations and all streaming infrastructure
- Consolidate shared parsing strategies into `TokenizerCore`
- Migrate all tests to use `ParsingTokenizer` as the sole production tokenizer
- Net reduction: **6,309 lines** across 16 deleted files

## Background & Context
- **Original Issue:** The Tokenizer module contained 6 separate implementations (`Tokenizer`, `EnhancedParsingTokenizer`, `FastParsingTokenizer`, `RefactoredParsingTokenizer`, `ParallelTokenizer`, `IncrementalTokenizer`) plus streaming infrastructure, but only `ParsingTokenizer` and `IncrementalTokenizer` were actually used in production.
- **Root Cause:** Incremental development added alternative implementations without removing prior ones, leading to ~5,000+ lines of dead code.
- **User Impact:** Increased cognitive load, longer build times, risk of divergent behavior between implementations, and maintenance burden for unused code.
- **Previous State:** 6 tokenizer implementations with overlapping logic (~600+ lines each), two conflicting "shared" implementation files (`SharedTokenizerImplementation.swift` and `TokenizerParsingStrategies.swift`), and streaming infrastructure used by no production code.

## Solution Approach
- **Core Solution:** Delete all unused tokenizer implementations and consolidate shared parsing logic into a single `TokenizerCore` module.
- **Technical Strategy:** Phase-by-phase deletion with test migration at each step, ensuring zero regressions.
- **Logic & Reasoning:** Only `ParsingTokenizer` is used by Parser, Interpreter, and DiagnosticsProvider. `IncrementalTokenizer` is kept for LSP/differential tokenization use cases.
- **Alternative Approaches:** Considered keeping `EnhancedParsingTokenizer` for its error recovery features, but decided on YAGNI — error recovery can be re-implemented when needed with a cleaner design.

## Implementation Details
- **Key Components:**
  - `TokenizerCore` (renamed from `SharedTokenizerImplementation`) — consolidated shared parsing logic
  - `ParsingTokenizer` — refactored to use `TokenizerCore.TokenData` directly, removing wrapper methods
  - `IncrementalTokenizer` — cleaned up dead `ChunkProcessor` reference

- **Files Deleted (16):**
  - Source: `Tokenizer.swift`, `EnhancedParsingTokenizer.swift`, `FastParsingTokenizer.swift`, `RefactoredParsingTokenizer.swift`, `ParallelTokenizer.swift`, `StreamingTokenizer.swift`, `SimpleTokenStream.swift`, `TokenStreamAdapter.swift`, `TokenizerParsingStrategies.swift`, `NumberParsingStrategies.swift`, `TokenizerResult.swift`, `EnhancedTokenizerError.swift`
  - Tests: `RefactoredParsingTokenizerTests.swift`, `FastParsingTokenizerTests.swift`, `EnhancedErrorHandlingTests.swift`, `StreamingTokenizerTests.swift`, `TokenStreamAdapterErrorHandlingTests.swift`, `TokenStreamTests.swift`

- **Files Modified (14):**
  - `TokenizerCore.swift` — absorbed methods from `TokenizerParsingStrategies` and `NumberParsingStrategies`
  - `ParsingTokenizer.swift` — removed private `TokenData`, uses `TokenizerCore.TokenData` directly
  - `IncrementalTokenizer.swift` — removed dead `ChunkProcessor` reference
  - `FeLangCore.swift` — updated doc comments from `Tokenizer` to `ParsingTokenizer`
  - Test files — migrated from `Tokenizer(input:)` to `ParsingTokenizer.tokenize()`

- **Behavioral Differences Handled:**
  - `Tokenizer` class parsed `-45` as a single negative literal; `ParsingTokenizer` parses as `minus` + `45` — tests adjusted accordingly
  - `Tokenizer` class emitted `.newline` tokens; `ParsingTokenizer` skips newlines — tests adjusted
  - Unicode normalization test removed (feature only existed in deleted `Tokenizer` class)

## Testing & Validation
- **Test Strategy:** All existing tests migrated to `ParsingTokenizer`; behavioral differences accounted for with test adjustments
- **Test Coverage:** 1,254 tests pass (down from 1,318 due to deleted test files for removed implementations)
- **Quality Assurance:** `swiftlint lint --fix && swiftlint lint && swift build && swift test` — all pass
- **Edge Cases:** Negative numbers, newline handling, Unicode normalization differences all handled

## Impact & Future Work
- **Breaking Changes:** None for production code (only `ParsingTokenizer` was used)
- **Performance Impact:** Faster build times due to fewer source files
- **Maintenance Impact:** Single tokenizer implementation to maintain instead of six
- **Future Enhancements:** E2E test helper extraction (identified but deferred to separate PR)
- **Technical Debt Resolved:** ~6,300 lines of dead code removed

🤖 Generated with [Claude Code](https://claude.com/claude-code)

---

<a href="https://app.devin.ai/review/fumiya-kume/felangkit/pull/385">
  <picture>
    <source media="(prefers-color-scheme: dark)" srcset="https://static.devin.ai/assets/gh-open-in-devin-review-dark.svg?v=1">
    Open with Devin
  </picture>
</a>

## Summary by CodeRabbit

* **リファクタリング**
  * トークナイザ処理を一本化し、ParsingTokenizerを標準エントリに統合。

* **改善**
  * 数値・文字列・コメントの基本解析が強化され、解析の一貫性と信頼性が向上。
  * 位置情報のオフセット計算がUnicodeスカラー基準に更新。

* **ドキュメント**
  * 使用例と公開API表記を ParsingTokenizer.tokenize 呼び出しに更新。

* **テスト**
  * テスト群を整理・再編し、ParsingTokenizer中心の検証に統一。

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: Sources/FeLangCore/Tokenizer/TokenizerCore.swift
[the remark this answers is not in the document]
Comment (fumiya-kume) [reply Sources/FeLangCore/Tokenizer/TokenizerCore.swift:1]: Fixed in `4447f58`. `parseDecimalNumber` now saves the index before consuming `e`/`E` and backtracks if no valid exponent digits follow. This ensures the number before the `e` is still returned as a valid token (e.g., `"1e"` → `integerLiteral("1")` + `identifier("e")`) instead of being silently dropped.

[the remark this answers is not in the document]
Comment (fumiya-kume) [reply Sources/FeLangCore/Tokenizer/TokenizerCore.swift:1]: Fixed in `4447f58`. All three base-specific parsers (`parseHexadecimalNumber`, `parseBinaryNumber`, `parseOctalNumber`) now save the index at entry and restore it before returning `nil`. Additionally, `parseNumber` now falls through to `parseDecimalNumber` when a base-specific parser fails, so `"0x"` produces `integerLiteral("0")` + `identifier("x")` instead of being silently consumed.

File: Sources/FeLangCore/Tokenizer/IncrementalTokenizer.swift
[the remark this answers is not in the document]
Comment (fumiya-kume) [reply Sources/FeLangCore/Tokenizer/IncrementalTokenizer.swift:1]: Fixed in `4447f58`. Changed `TokenizerUtilities.sourcePosition()` to use `input.unicodeScalars.distance(from:to:)` instead of `input.distance(from:to:)`, aligning the offset computation with the `SourcePosition.offset` documentation which specifies Unicode scalar count.

File: Sources/FeLangCore/Tokenizer/TokenizerCore.swift [hunk 1]
```diff
@@ -436,6 +434,239 @@ public enum SharedTokenizerImplementation {
...
+            }
+        }
+
+        // Return nil if unterminated (let caller handle the error)
+        guard foundClosing else { return nil }
+
+        let lexeme = String(input[start..<index])
+        let tokenType = TokenizerUtilities.stringLiteralTokenType(content: content)
+        return TokenData(type: tokenType, lexeme: lexeme)
+    }
```
[the remark this answers is not in the document]
Comment (fumiya-kume) [reply Sources/FeLangCore/Tokenizer/TokenizerCore.swift:582]: Fixed in `37bfea9`. Added 23 unit tests covering all 5 newly public APIs: `parseBasicNumber` (integer, decimal, leading dot, invalid), `parseBasicString` (double/single quote, escapes, unterminated, unknown escape), `parseComment` (single-line, multi-line, unterminated, not a comment), `isValidTokenBoundary` (end of input, before space, before identifier), and `canStartToken` (letter, digit, dot, quote, slash, space, newline).

File: Sources/FeLangCore/Tokenizer/TokenizerCore.swift [hunk 2]
```diff
@@ -436,6 +434,239 @@ public enum SharedTokenizerImplementation {
...
+                        index = input.index(after: afterStar)
+                        break
+                    }
+                }
+                index = input.index(after: index)
+            }
+
+            let lexeme = String(input[start..<index])
+            return TokenData(type: .comment, lexeme: lexeme)
+        }
```
[the remark this answers is not in the document]
Comment (fumiya-kume) [reply Sources/FeLangCore/Tokenizer/TokenizerCore.swift:635]: Fixed in `37bfea9`. Added `foundTerminator` flag to `parseComment`. When `*/` is not found before end of input, the index is reset to the start position and `nil` is returned, allowing the caller to handle the error appropriately.

File: Tests/FeLangCoreTests/Tokenizer/EnhancedNumberTests.swift [hunk 3]
```diff
@@ -49,21 +47,33 @@ struct EnhancedNumberTests {
...
+        // "1e5e" => "1e5" is a valid realLiteral, then "e" is an identifier
+        do {
+            let tokens = try ParsingTokenizer.tokenize("1e5e")
+            #expect(tokens.count == 3) // realLiteral('1e5') + identifier('e') + eof
+            #expect(tokens[0].type == .realLiteral)
+            #expect(tokens[0].lexeme == "1e5")
+            #expect(tokens[1].type == .identifier)
+            #expect(tokens[1].lexeme == "e")
         }
     }
```
[the remark this answers is not in the document]
Comment (fumiya-kume) [reply Tests/FeLangCoreTests/Tokenizer/EnhancedNumberTests.swift:114]: Addressed in `4447f58` (previous commit). The tokenizer now backtracks properly on invalid number prefixes instead of silently consuming them. For example, `"1e"` now produces `integerLiteral("1")` + `identifier("e")` and `"0x"` produces `integerLiteral("0")` + `identifier("x")`. All tests updated to reflect the correct behavior.

File: Sources/FeLangCore/Tokenizer/TokenizerCore.swift
[the remark this answers is not in the document]
Comment (fumiya-kume) [reply Sources/FeLangCore/Tokenizer/TokenizerCore.swift:1]: Fixed in `37bfea9`. Added optional `startIndex` parameter (defaults to `input.startIndex`) to both `parseNumberWithValidation` and `parseStringLiteral`. All error paths now compute real positions via `TokenizerUtilities.sourcePosition(from:startIndex:currentIndex:)` instead of using hardcoded `SourcePosition(line: 1, column: 1, offset: 0)`.

File: Sources/FeLangCore/Tokenizer/TokenizerCore.swift [hunk 4]
```diff
@@ -436,6 +429,239 @@ public enum SharedTokenizerImplementation {
...
```
[the remark this answers is not in the document]
Comment (fumiya-kume) [reply Sources/FeLangCore/Tokenizer/TokenizerCore.swift:635]: Fixed in `37bfea9`. `parseComment` now tracks whether a terminator was found. For unterminated block comments, the index is reset to the start position and `nil` is returned, consistent with the documented behavior ("skips over complete comments"). A new test `testParseCommentMultiLineUnterminated` verifies this behavior.

File: Sources/FeLangCore/Tokenizer/TokenizerCore.swift [hunk 5]
```diff
@@ -436,6 +429,239 @@ public enum SharedTokenizerImplementation {
...
+                    case "\"":
+                        content.append("\"")
+                        index = input.index(after: nextIndex)
+                    case "'":
+                        content.append("'")
+                        index = input.index(after: nextIndex)
+                    default:
+                        content.append(nextChar)
+                        index = input.index(after: nextIndex)
+                    }
```
[the remark this answers is not in the document]
Comment (fumiya-kume) [reply Sources/FeLangCore/Tokenizer/TokenizerCore.swift:566]: Fixed in `37bfea9`. The `default` case in `parseBasicString`'s escape sequence handler now resets the index and returns `nil` instead of silently appending the unknown escape character. This aligns with `parseStringLiteral`'s strict validation behavior. A new test `testParseBasicStringUnknownEscape` verifies that unknown escapes like `\x` are rejected.
